sources/packages/api/src/services/application/application.service.ts

```diff
@@ -1003,4 +1007,94 @@ export class ApplicationService extends RecordDataModelService<Application> {
+      throw new CustomNamedError(
+        `Student Application is not in the expected status. The application must be in application status '${ApplicationStatus.assessment}' to an MSFAA number be assigned.`,
+        INVALID_OPERATION_IN_THE_CURRENT_STATUS,
+      );
+    }
+
+    let msfaaNumberId: number;
+
+    // Checks if there is an MSFAA that could be considered valid.
+    const existingValidMSFAANumber =
+      await this.msfaaNumberService.getCurrentValidMSFAANumber(
+        application.studentId,
+      );
+    if (existingValidMSFAANumber) {
+      // Reuse the MSFAA that is still valid and avoid creating a new one.
+      msfaaNumberId = existingValidMSFAANumber.id;
+    } else {
+      // Get previously completed and signed application for the student
+      // to determine if an existing MSFAA is still valid.
+      const previousSignedApplication =
+        await this.getPreviouslySignedApplication(application.studentId);
+
+      // checks if the MSFAA number is still valid.
+      const hasValidMSFAANumber = this.msfaaNumberService.isMSFAANumberValid(
+        // Previously signed and completed application offering end date in considered the start date.
+        previousSignedApplication?.offering.studyEndDate,
+        // Start date of the offering of the current application is considered the end date.
+        application.offering.studyStartDate,
+      );
+
+      if (hasValidMSFAANumber) {
+        // Reuse the MSFAA number.
+        msfaaNumberId = previousSignedApplication.msfaaNumber.id;
+      } else {
+        // Create a new MSFAA number case the previous one is no longer valid.
+        const newMSFAANumber = await this.msfaaNumberService.createMSFAANumber(
+          application.studentId,
+        );
+        msfaaNumberId = newMSFAANumber.id;
+      }
+    }
+
+    // Associate the MSFAA number with the application.
+    application.msfaaNumber = { id: msfaaNumberId } as MSFAANumber;
+
+    return this.repo.save(application);
+  }
+
+  /**
+   * Gets the application that has an MSFAA signed date.
+   * @param studentId student id to filter the applications.
+   * @returns previous signed application if exists, otherwise null.
+   */
+  async getPreviouslySignedApplication(
+    studentId: number,
+  ): Promise<Application> {
+    return this.repo
+      .createQueryBuilder("applications")
+      .select(["applications.id", "msfaaNumbers.id", "offerings.studyEndDate"])
+      .innerJoin("applications.offering", "offerings")
+      .innerJoin("applications.msfaaNumber", "msfaaNumbers")
+      .innerJoin("applications.student", "students")
+      .where("applications.applicationStatus = :completedStatus", {
+        completedStatus: ApplicationStatus.completed,
+      })
+      .andWhere("students.id = :studentId", { studentId })
+      .andWhere("msfaaNumbers.dateSigned is not null")
+      .orderBy("offerings.studyEndDate", "DESC")
+      .limit(1)
```

Comment:
do we need `.limit(1)` and `getOne()` both?

Reply:
In terms of results, yes, but the limit(1) will return from DB only one record, that is what is needed.
I tested with an without the limit, as per below image, and surprisingly the limit is not added just because of the "getOne"
![image](https://user-images.githubusercontent.com/61259237/132914979-248281db-e479-4e18-b87b-4f496b76873b.png)